services/web/server/src/simcore_service_webserver/api_keys/_repository.py

```diff
@@ -0,0 +1,222 @@
+            )
+            .on_conflict_do_update(
+                index_elements=["user_id", "display_name"],
+                set_={
+                    "product_name": product_name
+                },  # dummy enable returning since on_conflict_do_nothing returns None
+                # NOTE: use this entry for reference counting in https://github.com/ITISFoundation/osparc-simcore/issues/5875
+            )
+            .returning(api_keys)
+        )
+
+        result = await conn.stream(insert_stmt)
+        row = await result.first()
+        assert row  # nosec
+
+        return ApiKey(
+            id=f"{row.id}",  # NOTE: remove after migration to str
+            display_name=row.display_name,
+            expiration=row.expires_at,
+            api_key=row.api_key,
+            api_secret=row.api_secret,
+        )
+
+
+async def list_api_keys(
+    app: web.Application,
+    connection: AsyncConnection | None = None,
+    *,
+    user_id: UserID,
+    product_name: ProductName,
+) -> list[ApiKey]:
+    async with transaction_context(get_asyncpg_engine(app), connection) as conn:
+        stmt = sa.select(api_keys.c.id, api_keys.c.display_name).where(
+            (api_keys.c.user_id == user_id) & (api_keys.c.product_name == product_name)
+        )
+
+        result = await conn.stream(stmt)
+        rows = [row async for row in result]
+
+        return [
+            ApiKey(
+                id=f"{row.id}",  # NOTE: remove after migration to str
+                display_name=row.display_name,
+            )
+            for row in rows
+        ]
+
+
+async def get_api_key(
+    app: web.Application,
+    connection: AsyncConnection | None = None,
+    *,
+    api_key_id: str,
+    user_id: UserID,
+    product_name: ProductName,
+) -> ApiKey | None:
+    async with transaction_context(get_asyncpg_engine(app), connection) as conn:
+        stmt = sa.select(api_keys).where(
+            (api_keys.c.id == int(api_key_id))  # NOTE: remove after migration to str
+            & (api_keys.c.user_id == user_id)
+            & (api_keys.c.product_name == product_name)
+        )
+
+        result = await conn.stream(stmt)
+        row = await result.first()
+
+        return (
+            ApiKey(
+                id=f"{row.id}",  # NOTE: remove after migration to str
+                display_name=row.display_name,
+                expiration=row.expires_at,
+                api_key=row.api_key,
+                api_secret=row.api_secret,
+            )
+            if row
+            else None
+        )
+
+
+async def delete_api_key(
+    app: web.Application,
+    connection: AsyncConnection | None = None,
+    *,
+    user_id: UserID,
+    product_name: ProductName,
+    api_key_id: str,
+) -> None:
+    async with transaction_context(get_asyncpg_engine(app), connection) as conn:
+        stmt = api_keys.delete().where(
+            (api_keys.c.id == int(api_key_id))  # NOTE: remove after migration to str
```

Comment:
Q: is there a test that will make this fail after the migration? That would be ideal. A note in the code "might" get forgotten.

Reply:
I created an issue for that and linked it to the code.